trackma/ui/cli.py

```diff
@@ -945,6 +969,26 @@ def _make_list(self, showlist):
         print('%d results' % len(showlist))
         print()
 
+    def _make_list_raw(self, showlist):
+        """
+        Helper function for printing a non-formatted show list
+        """
+        altnames = self.engine.altnames()
+
+        # List shows
+        for index, show in showlist:
+            if self.engine.mediainfo['has_progress']:
+                episodes_str = "{0}\t{1}".format(
+                    show['my_progress'], show['total'] or '?')
+            else:
+                episodes_str = "-"
+
+            # Get title (and alt. title) and if need be, truncate it
+            title_str = show['title']
+            if altnames.get(show['id']):
+                title_str += " [{}]".format(altnames.get(show['id']))
+
+            print(index,title_str, episodes_str,show['my_score'],sep='\t')
```

Comment:
Is the `index` actually something that tools need if they also have the name? I was generally under the impression that the index is primarily useful as a shortcut in interactive usage, but that's for humans and not for tools as they can just copy the proper anime name around. Besides, the index can change easily by any entry being added or removed from the list, making them very volatile and not safe to use for tooling. The unique `name` attribute of an entry should be superior in every way, aside from changing the language preference (english vs romaji for example), but the index would be even less safe then since the entire list would be have been reordered.

Similarly, what is the `name` command supposed to be used for?

And finally, I worry about the extensibility of this command as it currently uses a fixed list of columns, but what if we every wanted to add more to it or remove a column? I mean, even this implementation already has a variable number of columns depending on `has_progress`. 

The solutions I see are

1. Add a header line that names the respective columns
2. Just output JSON instead.

The decision for whether a tab-delimited table or JSON would be easier to use heavily depends on the use case & tooling, but JSON in general is much more flexible for tool-to-tool communication with rich data and frequently used by other tools.

Reply:
Well, `index` is indeed not needed.
Consider the following json approach:
```python
    def _make_list_raw(self, showlist):
        """
        helper function for printing a json formatted show list
        """
        altnames = self.engine.altnames()

        episode_str_current = ""
        episode_str_last = ""
        # list shows
        for index, show in showlist:
            if self.engine.mediainfo['has_progress']:
                episode_str_current = "{0}".format(show['my_progress'] or '0')
                episode_str_last = "{0}".format(show['total'] or '?')

            # get title (and alt. title) and if need be, truncate it
            title_str = show['title']
            if altnames.get(show['id']):
                title_str += " [{}]".format(altnames.get(show['id']))

            # ensure score is string; anilist can have string scores such as stars or smiles
            score = "{0}".format(show['my_score'])

            # print(index,title_str,episodes_str,show['my_score'],sep='\t')
            j = {
                "title": title_str,
                "current_episode": episode_str_current,
                "final_episode": episode_str_last,
                "score": score
            }
            print(json.dumps(j))

```
which prints:
```
{"title": "Ars no Kyojuu", "current_episode": "1", "final_episode": "12", "score": "0"}
{"title": "Blue Lock", "current_episode": "13", "final_episode": "24", "score": "5.5"}

```
As current_episode is number and final_episode could be string (?), I format both.
Score can also potentially be stars or smiles on anilist, so I am not sure how to deal with that there.

`trackma list_raw | jq -r '.title'` works as expected on my end. Of course, score, final_episode and current_episode also work.

---

As mentioned before `trackma info` can be a bit slow as it has to print more info. `name` would just print the name for a given index. But, you are right the index can change, so is better to ignore it. As I said above feel free to ignore the `name`.

If this json approach is suitable, I can make another PR or change this one.

Reply:
After making some changes to my personal adl script, to use trackma list_json, I also propose the following change to my above code, in order to also print the color code.
```python
            # json dictionary
            j = {
                "title": title_str,
                "current_episode": episode_str_current,
                "final_episode": episode_str_last,
                "score": score
            }

            # Color title according to status
            if show['status'] == utils.Status.AIRING:
                estimate = utils.estimate_aired_episodes(show)
                if estimate and show['my_progress'] < estimate:
                    # User is behind the (estimated) aired episode
                    j["color"] = _COLOR_BEHIND
                else:
                    j["color"] = _COLOR_AIRING
            else:
                j["color"] = _COLOR_RESET

            print(json.dumps(j))
```
This prints:
```
{"title": "Ars no Kyojuu", "current_episode": "1", "final_episode": "12", "score": "0", "color": "\u001b[0;34m"}
{"title": "Blue Lock", "current_episode": "13", "final_episode": "24", "score": "5.5", "color": "\u001b[0;34m"}
```
The else with color_reset could be left out, but perhaps this might cause issues in other people's code.
`trackma list_json | jq -r '.color // "color value"'` could be used in bash.

Reply:
~~`I suggest using an enum with string values istead of ANSI escape codes for "airing/behind" status. The color representation of a status is an implementation detail.~~

Actually, I'd put "airing" in a field on its own and the information used for the "behind" field, the estimated latest episode, also in its own field because then tooling can decide for itself how it wants to treat entries like that.

Going forward, we might also be interested in an option that selects the fields to be printed, but that can come in later as it would be backwards-compatible.

Reply:
Good idea. I made changes for this.

```python
            # json dictionary
            j = {
                "title": title_str,
                "current_episode": episode_str_current,
                "total_episodes": episode_str_last,
                "score": score,
                "airing_status": False,
                "behind_status": False
            }

            # Check if show is airing and if user is behind
            if show['status'] == utils.Status.AIRING:
                estimate = utils.estimate_aired_episodes(show)
                if estimate and show['my_progress'] < estimate:
                    # User is behind the (estimated) aired episode
                    j["airing_status"] = True;
                    j["behind_status"] = True;
                else:
                    j["airing_status"] = True;

            print(json.dumps(j))
```
I have tested on my end and prints as expected.
```json
{"title": "Saikyou Onmyouji no Isekai Tenseiki", "current_episode": "1", "total_episodes": "?", "score": "0", "airing_status": true, "behind_status": false}
{"title": "Scrapped Princess", "current_episode": "2", "total_episodes": "24", "score": "0", "airing_status": false, "behind_status": false}
{"title": "Sugar Apple Fairy Tale", "current_episode": "1", "total_episodes": "12", "score": "0", "airing_status": true, "behind_status": true}
{"title": "TRIGUN STAMPEDE", "current_episode": "1", "total_episodes": "?", "score": "0", "airing_status": true, "behind_status": false}
```

Reply:
:+1: Let's close this in favor of more focues discussions.